### PR TITLE
[Generate] correct encoder_outputs are passed without attention_mask

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

Very very edge case is solved in this PR which occurs if `"encoder_outputs"` is passed in combination without passing `"attention_mask"`, but a model that accepts an `"attention_mask"`.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: tests/test_generation_utils.py [hunk 1]
```diff
@@ -1887,3 +1887,25 @@ def test_generate_pixel_values_as_encoder_kwarg(self):
 
         self.assertListEqual(output_sequences.tolist(), output_sequences_kwargs.tolist())
         self.assertEqual(output_sequences.shape, (2, 5))
+
+    def test_generate_encoder_outputs_attention_mask(self):
+        # article = """Justin Timberlake and Jessica Biel, welcome to parenthood."""
```
Comment (patrickvonplaten) [inline tests/test_generation_utils.py:1892]:
```suggestion
```

File: tests/test_generation_utils.py [hunk 2]
```diff
@@ -1887,3 +1887,25 @@ def test_generate_pixel_values_as_encoder_kwarg(self):
...
+        # tokenizer = BartTokenizer.from_pretrained("hf-internal-testing/tiny-random-bart")
```
Comment (patrickvonplaten) [inline tests/test_generation_utils.py:1893]:
```suggestion
```

File: tests/test_generation_utils.py [hunk 3]
```diff
@@ -1887,3 +1887,25 @@ def test_generate_pixel_values_as_encoder_kwarg(self):
...
+        # model = BartForConditionalGeneration.from_pretrained("hf-internal-testing/tiny-random-bart", max_length=5).to(
```
Comment (patrickvonplaten) [inline tests/test_generation_utils.py:1894]:
```suggestion
```

File: tests/test_generation_utils.py [hunk 4]
```diff
@@ -1887,3 +1887,25 @@ def test_generate_pixel_values_as_encoder_kwarg(self):
...
+        #     torch_device
```
Comment (patrickvonplaten) [inline tests/test_generation_utils.py:1895]:
```suggestion
```

File: tests/test_generation_utils.py [hunk 5]
```diff
@@ -1887,3 +1887,25 @@ def test_generate_pixel_values_as_encoder_kwarg(self):
...
+        # )
```
Comment (patrickvonplaten) [inline tests/test_generation_utils.py:1896]:
```suggestion
```

File: tests/test_generation_utils.py [hunk 6]
```diff
@@ -1887,3 +1887,20 @@ def test_generate_pixel_values_as_encoder_kwarg(self):
...
+        model = model.to(torch_device)
+
+        encoder = model.get_encoder()
+
+        encoder_outputs = encoder(input_values)
+
+        output_sequences_no_mask = model.generate(encoder_outputs=encoder_outputs).cpu()
+        output_sequences_with_mask = model.generate(
+            encoder_outputs=encoder_outputs, attention_mask=attention_mask
+        ).cpu()
```
Comment (sgugger) [inline tests/test_generation_utils.py:1904]: As usual, when formatted like this by black, I'd go for two lines :-)

**Review by sgugger — APPROVED** [review]
Thanks for fixing!

File: tests/test_pipelines_automatic_speech_recognition.py [hunk 7]
```diff
@@ -215,3 +215,20 @@ def test_xls_r_from_en(self):
         filename = ds[40]["file"]
         output = speech_recognizer(filename)
         self.assertEqual(output, {"text": "Ein Mann sagte zu dem Universum, Sir, ich bin da."})
+
+    @slow
+    @require_torch
+    @require_torchaudio
+    def test_wav2vec2_2_bart(self):
```
Comment (Narsil) [inline tests/test_pipelines_automatic_speech_recognition.py:222]: Is this test here intentional, using wav2vec2 (not generative) leads me to believe it isn't intentional.

Regardless, `pipeline(model="patrickvonplaten/wav2vec2-2-bart-base")` should work out of the box FYI.

File: tests/test_pipelines_automatic_speech_recognition.py [hunk 8]
```diff
@@ -215,3 +215,20 @@ def test_xls_r_from_en(self):
...
+        speech_recognizer = pipeline(
+            task="automatic-speech-recognition",
+            model="patrickvonplaten/wav2vec2-2-bart-base",
+            feature_extractor="patrickvonplaten/wav2vec2-2-bart-base",
+            tokenizer=AutoTokenizer.from_pretrained("patrickvonplaten/wav2vec2-2-bart-base"),
+            framework="pt",
+        )
+
+        ds = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation").sort("id")
```
Comment (Narsil) [inline tests/test_pipelines_automatic_speech_recognition.py:231]: Is the sort necessary ?

**Review by Narsil — APPROVED** [review]
LGTM

The wav2vec2 doesn't seem to belong, if it does maybe a test should explain what the test is actually about.

File: tests/test_pipelines_automatic_speech_recognition.py (hunk 8 above)
Comment (patrickvonplaten) [reply tests/test_pipelines_automatic_speech_recognition.py:231]: yeah since it different versions datasets might break the order

Comment (Narsil) [reply tests/test_pipelines_automatic_speech_recognition.py:231]: This seems like a very odd thing in `datasets` that order could be messed up by subsequent versions...

Comment (patrickvonplaten) [reply tests/test_pipelines_automatic_speech_recognition.py:231]: @anton-l and I have experienced this now a couple of times so tat we'll just make sure this way

Comment (Narsil) [reply tests/test_pipelines_automatic_speech_recognition.py:231]: Good to know !

**Pull request merged**